### ISSUE TITLE
getState and setState

### ISSUE THREAD

**Issue by zimt28** [CodeAdventure/space-ui#14]
I'm a bit lost again :) I have no idea why this code doesn't work now, it aready did on my other branch -.-

Some code:
```coffeescript
class App.CasesStore extends Space.ui.Store

  Dependencies:
    cases: 'Cases'
    actions: 'Actions'

  setInitialState: ->

    limit = 5

    {
      case: null
      cases: @cases.find currentStatus: { $lte: limit }
    }

  configure: ->

    @listenTo(
      @actions.setCurrentCase, @_setCurrentCase
    )

  _setCurrentCase: (caseId) ->
    @setState 'case', @cases.findOne caseId
```

When I log `@getState 'case'` in the `_setCurrentCase` function, I get the **whole** (why?) object from `setInitialState`, with the correct case document on the  `case` property.

```coffeescript
class App.CasesShowController

  Dependencies:
    actions: 'Actions'
    tracker: 'Tracker'
    router: 'Router'

  onDependenciesReady: ->

    self = this

    @router.route 'cases.show',
      path: 'show/:_id'

      waitOn: ->
        singleSubs.subscribe 'case', @params._id

      onBeforeAction: ->
        caseId = @params._id

        self.tracker.nonreactive -> self._setCurrentCase caseId

  _setCurrentCase: (caseId) ->
    @actions.setCurrentCase caseId
```

This seems to work, when I log the `caseId`, I get the right _id.

```coffeescript
class App.CasesShowMediator extends Space.ui.Mediator

  @Template: 'CasesShow'

  Dependencies:
    store: 'CasesStore'

  templateHelpers: ->

    state: => {
      case: @store.getState().case
    }
```

This is strange. When I go into the `templateHelpers` function and log `@store.getState().case`, I get `null`, which is the initial value. But in the store's `_setCurrentCase` function, it logged the state object with the correct case (although I would have expected only the case).

As I already said, the code worked on another branch, with an older version of space:ui, two days ago. I think `getState` and `setState` do weird stuff. Or did I miss some changes? Or is it because of a wrong order (I don't think that's really possible)?

Thank you :)

-- Additional question: Are states reactive?

**Comment by DominikGuzei** [comment]
Hey @zimt28 :-)

Currently `getState` always returns the whole state object and doesn't take any arguments. But I like what you tried to do (call it with a subpath like `setState`) so I think it would be good to add this to the API. For now you have to use it like in the `templateHelpers` -> `getState().case`

Yes, the `state` property of stores is just a simple Meteor [ReactiveVar](http://docs.meteor.com/#/full/reactivevar), so any function that asks a store for the state becomes a reactive dependency of it, which means that the dependent function (the place where you call `@store.getState()` is re-run every time the state changes.

**Comment by DominikGuzei** [comment]
To your problem with not getting the state: it's strange because I can't see any difference to the code used in the TodoMVC example … I will take a look and tell you if I find anything strange.

For your interest, these "problems" could be caused by the new `Space.Injector` I introduces in `space:base@1.2` which is a complete and simplified rewrite of `Dependance.Injector`. It could be that there are still some edge cases that cause these bugs. I will try to fix these issues as quickly as possible, so that you can focus on building your app again ;-)

**Comment by zimt28** [comment]
Thanks for answering! I took the `getState` code from the [example](https://github.com/CodeAdventure/space-ui/blob/master/examples/TodoMVC/client/stores/todos_store.coffee#L48). It may not be so important if it's `@getState().case` or `@getState 'case'`, the version with the string just looks a bit more like `@setState`, so that might be a nice addition to keep the code consistent :) And by making the argument optional, the two possibilities could even be combined :+1:

**Comment by zimt28** [comment]
And I can't find any difference as well, as it did already work the only difference now might be the new space:ui or space:base. I'm curious about it :)

**Comment by DominikGuzei** [comment]
Haha, nice – you just found an if statement that will never be `true` because `@getState('activeFilter')` will always return the whole `state`. This is what you get, when you don't write tests!

**Comment by zimt28** [comment]
Yay :D Maybe [this line](https://github.com/CodeAdventure/space-ui/blob/master/source/flux/store.coffee#L11) could be changed to 
```coffeescript
getState: (property) ->
  if property
  	@state.get()[property]
  else
  	@state.get()
```
Would need to try this but I think it should work. I still need to get into testing myself, so that's ok ;-)

**Comment by DominikGuzei** [comment]
I think I found at least one part of the confusion you described:
> When I go into the templateHelpers function and log @store.getState().case, I get null, which is the initial value

This is expected, because `templateHelpers` only is executed once, when the mediator is created! Only the functions that really access the store instance are re-run:

```coffeescript
templateHelpers: ->
  # this scope is only executed once
  state: => 
      # here you always get the latest changes!
      console.log @store.getState().case
      case: @store.getState().case
```

**Comment by zimt28** [comment]
She state should be reactive so it should work, shouldn't it? At least yesterday it did work great :(

**Comment by DominikGuzei** [comment]
Yeah the state is reactive, I just added the same kind of code you posted to the TodoMVC example and it worked perfectly! Its really strange that it doesnt work in your case. Can you upload this code to some repository or send me a zip file? Would make it easier to find your problem

**Comment by zimt28** [comment]
I just noticed, that it does actually work. It just does not work for the first `case/:_id` page I try to access. And when I update the document, the state function won't be called again.

**Comment by DominikGuzei** [comment]
> And when I update the document, the state function won't be called again.
This is also expected, since `findOne` is not reactive! You have to use `find` instead if you want to it to be reactive ;-)

**Comment by zimt28** [comment]
I'm pretty sure it is reactive ..
```javascript
Template.CasesShow.helpers({
  case: function () {
    return Cases.findOne(Router.current().params._id);
  }
});
```
This works. And the [docs for findOne](http://docs.meteor.com/#/full/findone) say:
```
reactive Boolean
(Client only) Default true; pass false to disable reactivity
```

**Comment by DominikGuzei** [comment]
Ah that's true :-) I might have remembered the usage at server.

**Comment by zimt28** [comment]
So what's wrong then?? :/ I don't like sharing my app as all the server access codes and stuff are included but maybe we can use TeamViewer and you might have a look?

**Comment by zimt28** [comment]
I would just text you on facebook

**Comment by DominikGuzei** [comment]
could you setup a small example app where I can look at the problem? I think that would be easier, than remote debugging ;-)

**Comment by zimt28** [comment]
I'll try

**Comment by DominikGuzei** [comment]
Awesome, thanks! :-)

**Comment by DominikGuzei** [comment]
@zimt28 I just released `space:ui@3.4.3` which adds tests for all the store `state` related methods. I also improved the API like we discusses, so you can now `store.setState('my.nested.property', value)` as well as get nested properties like `store.getState('my.nested.property')`. Additionally there are two simpler aliases now: `store.get('my.nested.property')` and `store.set('my.nested.property', value)` ;-)

**Comment by zimt28** [comment]
Here we go :) See https://github.com/zimt28/space-ui-problem

**Comment by DominikGuzei** [comment]
Cool, I will check it out! I also found another edge case where the new injector doesn't work as expected ;-) it will be some more work, but hopefully we have ironed out these problems soon!

**Comment by zimt28** [comment]
Updated the demo again. You can uncomment some code in `client/views/cases_show/cases_show.html`, I just want to show the difference to a regular helper. Maybe it's something about the router as well? Just because the helper does some strange things as well (see comment in file). Note that im using `iron:router@0.9.4` as another important package is still getting updated to support iron:router v1.

**Comment by DominikGuzei** [comment]
Ok great, in the meantime I already discovered a serious bug in the `@setState` method. I will fix that and see if it helps.

**Comment by zimt28** [comment]
One more commit, it's not a problem with iron:router: https://github.com/zimt28/space-ui-problem/commit/d43594f2d9af907901bded4b91dcdcbe3d4eb22f

**Comment by DominikGuzei** [comment]
Hey @zimt28 I solved all of your case problems:

1. Was a bug in `Space.Injector` which is solved now in `space:ui@3.4.4`
2. Inside the case (updating issue) is actually a problem in your code, but you have to understand Meteor reactivity quite well: Reactivity only works within a `Tracker.Computation`, and this is **not** automatically done everywhere. The template helpers actually are one of the only places where you can just call `findOne` and have it be reactive. So your problem here is, that you assign the return value of `findOne` in the `_setCurrentCase` store method to `state.case`. But `_setCurrentCase` is not enclosed in a `Tracker.Computation` (e.g: via Tracker.autorun ->) so this is not reactive! At the same time, `findOne` just returns a simple plain object with normal javascript properties, there is no magic! That's the reason why `find` is different: It returns a `Cursor` of a collection and cursors handle the reactivity and access for you.
3. The last issue (Case 4) is also a problem with your code. Refresh doesn't work because you use an anonymous collection (client side only / not persisted) in the `App.Application` which always generates new ids for the cases on every refresh ;-) So this could be solved by making it a real collection instead.

Bottom line for issue 2. is that you three choices:
1. You could use `find` instead of `findOne` and everything should work reactively
2. Or you have to send an action to the store to update the title, which would change the `state` and thus also be reactive.
3. You wrap your call to `findOne` within `_setCurrentCase` into a `Tracker.autorun` block to make it reactive. However, you have to remember that `findOne` is only reactive client-side, so it doesn't pickup changes from the server.

The first solution is more flexible when you want your views to always be reactive, no matter where the collection was updated. The others are kind of fragile because `findOne` can be a bit confusing at times.

**Comment by zimt28** [comment]
Ohh, great!! It works now :) I'll play around with the reactivity later, the last issue is clear .. Totally forgot that I just used a client-side collection as it was enough to reproduce my issues :p

**Comment by DominikGuzei** [comment]
Awesome, thank you for the great reporting and help to fix these issues!

**Comment by zimt28** [comment]
Thank you, too! I've ported another page to space:ui and I'm amazed how clean and structured my code gets - and it saved my a lot of code as there are better ways to do things now :D Love it <3

Maybe you can help me beautify this code a bit. I have three states that I want to give value.

```coffeescript
  setInitialState: ->
    {
      case: null
      cases: [] # this
      activeCases: [] # this
      archivedCases: [] # and this
      archiveLimit: null
    }
```

This is from my controller:
```coffeescript
@tracker.nonreactive =>
  @actions.setArchiveLimit limit
  @actions.setupCases()
  @actions.getActiveCases()
```

Now the ugly part, the store:
```coffeescript
  _setupCases: ->
    @setState 'activeCases', @cases.find { currentStatus: { $lte: @getState('archiveLimit') } }, sort: { createdAt: -1 }
    @setState 'archivedCases', @cases.find { currentStatus: { $gt: @getState('archiveLimit') } }, sort: { createdAt: -1 }

  _getActiveCases: ->
    @setState 'cases', @cases.find { currentStatus: { $lte: @getState('archiveLimit') } }, sort: { createdAt: -1 }

  _getArchivedCases: (state) ->
    @setState 'cases', @cases.find { currentStatus: { $gt: @getState('archiveLimit') } }, sort: { createdAt: -1 }
```

As you can see, I have to repeat the mongo stuff which is ugly. I tried something like this, but it didn't work :(

```coffeescript
@tracker.nonreactive =>
  @actions.setArchiveLimit limit
  @actions.setupCases()
```

```coffeescript
  _setupCases: ->
    @_getActiveCases()
    @_getActiveCases('activeCases')
    @_getArchivedCases('archivedCases')

  _getActiveCases: (state) ->
    unless state? then state = 'cases'
    @setState state, @cases.find { currentStatus: { $lte: @getState('archiveLimit') } }, sort: { createdAt: -1 }

  _getArchivedCases: (state) ->
    unless state? then state = 'cases'
    @setState 'cases', @cases.find { currentStatus: { $gt: @getState('archiveLimit') } }, sort: { createdAt: -1 }
```

Any idea how I get make this a bit nicer (and keep it working)? Thanks :)

**Comment by DominikGuzei** [comment]
As I see it, you have two queries that depend on the value of `archiveLimit`.

first, I think you could simplify your code like this:
```coffeescript
_setupCases: ->
    @setState 'activeCases', @cases.find { currentStatus: { $lte: @getState('archiveLimit') } }, sort: { createdAt: -1 }
    @setState 'archivedCases', @cases.find { currentStatus: { $gt: @getState('archiveLimit') } }, sort: { createdAt: -1 }

  _getActiveCases: -> @setState 'cases', @get('activeCases')

  _getArchivedCases: (state) -> @setState 'cases', @get('archivedCases')
```

if you really want `activeCases` and `archivedCases` to be fully reactive (change whenever `archiveLimit` changes) you would have to setup a `Tracker.Computation` via `Tracker.autorun` etc. but that's probably not much nicer.

Computed properties are actually an interesting problem that Meteor doesn't (elegantly) really solve. However one could also argue that the `store` should actually not be concerned about the current view, so it might be even cleaner to move this switching logic into the mediator. Having a `currentCases` state property on the mediator that either points to the `activeCases` or `archivedCases` … so many possiblities :-D

**Comment by DominikGuzei** [comment]
Hey @zimt28, I am happy to tell you that `setInitialState` is now also reactive in `Space.ui.Mediator` and `Space.ui.Store` so you can use `Mongo.Collection#findOne` there and it will reactively update as expected :wink: This was introduced in version "space:ui@4.3.1"
